### ISSUE TITLE
UnknownFortranDataTypes

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#2441]
I have found a few code examples in LFRic where an UnknownFortranDataType is used:
```
  real(kind=r_tran), allocatable, dimension(:) :: dep_pts_z_data_post
  ! UnknownFortranDataType
  INTEGER(KIND = i_def), PUBLIC, PROTECTED :: vertical_limit = emdi
  ! UnknownFortranDataType
  CHARACTER(LEN = str_long), PUBLIC :: log_scratch_space
```
Now some of these creates problem with the driver creation (since it can't find the symbols in the modules, so it creates them as local variables).

Two main issues:
1. Is this a known issue, or should I verify these problems and open separate tickets?
2. It would be really helpful if we could add a comment (like in the example above) that a data type is unkown. It just means that when you get e.g. compilation problems (e.g. because public etc declarations make no sense for variables in a program), you can immediately see where this strange declaration comes from.

I have a rather simple patch (not thoroughly tested, but it created the code above):
```
.../psyclone/src/psyclone$ git diff  psyir/backend/fortran.py
diff --git a/src/psyclone/psyir/backend/fortran.py b/src/psyclone/psyir/backend/fortran.py
index 3c6db99db..64abbd41a 100644
--- a/src/psyclone/psyir/backend/fortran.py
+++ b/src/psyclone/psyir/backend/fortran.py
@@ -557,7 +557,9 @@ class FortranWriter(LanguageWriter):
                     return f"{self._nindent}{decln}\n"
 
                 decln = symbol.datatype.declaration
-                return f"{self._nindent}{decln}\n"
+                return (f"{self._nindent}{self._COMMENT_PREFIX} "
+                        f"UnknownFortranDataType\n"
+                        f"{self._nindent}{decln}\n")
             # The Fortran backend only handles unknown *Fortran* declarations.
             raise VisitorError(
                 f"{type(symbol).__name__} '{symbol.name}' is of "
```

**Comment by sergisiso** [comment]
> Is this a known issue, or should I verify these problems and open separate tickets?

This needs to be fixed. But how does it know the declaration if it doesn't found them on the module, is something that we manually construct? If they appear locally seems to indicate that we don't construct or not copy its interface, which is the think that tells that it comes from a import. 

> It would be really helpful if we could add a comment (like in the example above) that a data type is unkown. It just means that when you get e.g. compilation problems (e.g. because public etc declarations make no sense for variables in a program), you can immediately see where this strange declaration comes from.

I am ok with the comment if its helpful. Should commentablemixin be more broad than just nodes? But ideally we should fix the original problem.

**Comment by hiker** [comment]
OK, my explanation was mostly wrong, but the problem does exist. Indeed, the symbol is in the symbol table. But for an output variable `log_scratch_space` from a module, the driver creates a local variable `log_scratch_space_post` of the same type as `log_scratch_space`.

So the symbol table contains `log_scratch_space` imported from `log_mod`, and it contains `log_scratch_space_post` as newly declared variable (that contains the expected output value of this variable), but it's an  UnknownFortranType containing:

`UnknownFortranType('CHARACTER(LEN = str_long), PUBLIC :: log_scratch_space')`

So, when the symbol table is used to write all variables, it first imports `log_scratch_space` from `log_mod` - correct. Then when it tries to declare `log_scratch_space_post`, it actually writes out the UnknownFortranType as above,  i.e. it adds another variable with the name `log_scratch_space`. Besides duplicating the name, the `public` statement is then also not allowed.

**Comment by hiker** [comment]
Yes, not immediately obvious, but I think so. What do you think of adding a comment to these declarations when written out? I would find it very useful to know when something is not handled properly by PSyIR (now that I have written that, maybe the same for code block)?

If you agree with only for Unknown Types, I'll leave this open and add a small PR. If you think it might also be useful for CodeBlocks, I might just open a new ticket. Or we just close it if you think the comment is useless (imho I wouldn't go for adding a comment mixing to symbols - it's easy enough to test in any python script if a type is an UnknownFortanType, I just want to it as an indication to a user that actually looks at the created code.

**Comment by sergisiso** [comment]
I am ok with the comments if they are useful when looking at the output.
I also think the CodeBlock comments will be useful and I was even thinking of leaving a comment for each loop not parallelised explaining the reason (but that would be specific to the transformation).

**Comment by arporter** [comment]
With the switch to the PSyIR backend has this issue gone away?
